### ISSUE TITLE
Fuzzy finder improvements

### ISSUE THREAD

**Issue by olafurpg** [sourcegraph/sourcegraph#21201]
Following up from a call with @limitedmage, opening this issue to track things that can be improved after merging #20921

## Blocking GA
- [x] Iron out design, pending proposal https://www.figma.com/file/ZYu2iiTp4AmnHDXf9EX2sY/File-fuzzy-find?node-id=1%3A6549&viewport=2392%2C586%2C1 (merged PR https://github.com/sourcegraph/sourcegraph/pull/21876 )
- [x] Optimize "download files" step on the backend. It currently takes 10-20 seconds in the chromium repo to download all files. Fixed in progress https://github.com/sourcegraph/sourcegraph/pull/21422
- [x] Be more case insensitive in smaller repos (<25k files). Fixed in #21361
- [x] Stop using cache storage. Pending fix https://github.com/sourcegraph/sourcegraph/pull/21550
- [x] Improve accessibility (PR https://github.com/sourcegraph/sourcegraph/pull/21919, #22054)
- [ ] Don't insert git SHA into link URLs when on the default branch (pending PR https://github.com/sourcegraph/sourcegraph/pull/22133)
- [ ] Fix the bug that causes files to get re-downloaded when going from repo view to blob view (easy to reproduce in chromium/chromium repo). (pending PR https://github.com/sourcegraph/sourcegraph/pull/22133)

## Non-blocking

- [x] Consider using different keyboard shortcut that is also active while the search bar is focused. (superseded by https://github.com/sourcegraph/sourcegraph/issues/22229 )
- [x] Make focusIndex state persistent across blob views (superseded by https://github.com/sourcegraph/sourcegraph/issues/22221 )
- [x] Make dark theme less dark (superseded by https://github.com/sourcegraph/sourcegraph/issues/22222)
- [x] (high priority) Remove `fileAnchor?.click()` DOM manipulation https://github.com/sourcegraph/sourcegraph/pull/20921/files#r637086203 (superseded by https://github.com/sourcegraph/sourcegraph/issues/22223)
- [x] Address non-blocking review comments from https://github.com/sourcegraph/sourcegraph/pull/21876#pullrequestreview-678690607
- [x] turn `renderFuzzyResult` method into a component, or remove `JSX.Element` logic. https://github.com/sourcegraph/sourcegraph/pull/20921/files#r637092526 (superseded by https://github.com/sourcegraph/sourcegraph/issues/22226 )
- [x] Remove `fuzzyFinder=debug` query parameter once `experimentalFeature.fuzzyFinder` is supported on dotcom
- [x] Use theme-aware CSS variables for `HighlightedText` yellow and black colors
- [x] Consider using `<mark>` for the highlighted section instead of `<span>` for highlighted result items
- [x] Mock graphql query to make it easier to test
~~- Use BEM notation in CSS~~
- [x] Use reach `<Dialogue` component instead of hand-written modal
- [x] Fine-tune bloom filter false positive ratio for ideal latency/memory-usage balance.
- [x] Remove usage of global eslint-disable 
- [x] Inline `offsetSum` at call site (wontfix)
- [x] Inline properties into `FuzzyModalProps`
- [x] Rename `CaseSensitiveFuzzySearch` to `WordSensitiveFuzzySearch` because "case insensitive" would allow the query "bar" to match "Foobar" whereas the current implementation does not match because it requires the query to start a word
- [x] Add storybook tests
- [x] Investigate why the query "attributes" in sourcegraph/sourcegraph doesn't return results for the filename ".gitattributes"

FYI @novoselrok

**Comment by malomarrec** [comment]
@olafurpg I'll bring this up in the search UX spring planning, and get this planned for.

**Comment by felixfbecker** [comment]
Per our [accessibility principles](https://docs.sourcegraph.com/dev/background-information/web/accessibility), accessibility should be added to the list above.

> People of all abilities should have first-class access to code and coding. In order to achieve this it is vital that we build our application to be accessible to all. All client code should strive to meet our accessibility standards.

Currently, if using a screen reader, the currently selected element is not announced when using the arrow keys, meaning visually impaired users can't know what item is selected.

The [`role="combobox"`](https://www.w3.org/TR/wai-aria-practices-1.2/#combobox) role probably matches this widget the closest (with a "popup" that is _always visible_) and the selected element would need to have `aria-selected="true"` applied. The keyboard interaction requirements (arrow keys, enter) should already be met.

**Comment by olafurpg** [comment]
@felixfbecker That's a good suggestion. I took a stab at improving the accessibility in the commit here https://github.com/sourcegraph/sourcegraph/commit/9d0cbfb714540886199aa9d36721f1c0ecad9fa0 I opened a PR with that commit here https://github.com/sourcegraph/sourcegraph/pull/21919 I'll update the PR to remove the unrelated commits once the PR https://github.com/sourcegraph/sourcegraph/pull/21876 is merged.

**Comment by olafurpg** [comment]
We just merged #21919, which I believe addressed all known accessibility issues.

@malomarrec Any suggestions for next steps towards unblocking GA?

**Comment by limitedmage** [comment]
IMO at this point we can turn the feature flag on for the Sourcegraph org on sourcegraph.com, post on #dev-announce, and let the team give it a try and send feedback :)

**Comment by felixfbecker** [comment]
The screen reader support isn't working yet (see https://github.com/sourcegraph/sourcegraph/pull/21919#issuecomment-859716845) but I agree we should enable it for our org :)

**Comment by olafurpg** [comment]
How do we enable the feature flag for our Sourcegraph org? (I thought it was already enabled 🤔 )

I just opened #22054 fixing screen reader support.

**Comment by tsenart** [comment]
Awesome feature :)

Some feedback from experimentation today.

1. The modal looks too wide for repos that don't have very deep directory hierarchies.

<img width="1576" alt="Screen Shot 2021-06-16 at 10 45 08" src="https://user-images.githubusercontent.com/67471/122188245-118d0880-ce90-11eb-8192-beba2f7cdbc5.png">

2. Click the Show more button doesn't seem to do anything? Or maybe it does, but there's no visual cue.

<img width="1576" alt="Screen Shot 2021-06-16 at 10 47 29" src="https://user-images.githubusercontent.com/67471/122188659-75afcc80-ce90-11eb-8b80-36b6f336ead8.png">

3. Would be awesome to show files in a [frecency](https://en.wikipedia.org/wiki/Frecency) order rather than lexical (perhaps a flag to enable that). This is a larger feature, I know :)

4. Inspired by @tjdevries telescope, I now kind of expect a fuzzy finder to show a preview of the file on the side. That'd be awesome too.

**Comment by ryanslade** [comment]
Very cool!

Would it be possible to cache the list on the client? For a large repo such as Chromium it needs to load the list every time I browse to a different area in the repo which takes can take a long time.

**Comment by olafurpg** [comment]
@ryanslade the original implementation actually cached the filenames but I removed this feature here https://github.com/sourcegraph/sourcegraph/pull/21550 The filenames are still cached within the React single-page app so you can navigate between multiple files without re-downloading the files. However, there seems to be a bug that causes the files to get re-downloaded when you go from the repo-view to the blob-view. This is only really a problem in XXL repos like Chromium, it's not that bad in other large repos with <100k filenames like torvalds/linux.

I added a TODO above to investigate this repo-view -> blob-view bug.

**Comment by olafurpg** [comment]
@tsenart thank you for the feedback!

> The modal looks too wide for repos that don't have very deep directory hierarchies.

@rrhyne Could someone from design take a look at this? Maybe we can reduce the width for repos with only a small number of short filenames 🤔 

> Click the Show more button doesn't seem to do anything? Or maybe it does, but there's no visual cue.

@tsenart originally, the "show more" button appeared at the bottom of the list when you scrolled down. More results appear when you click on it. If you're scrolled at the top, the only visual hint is the "600+ results" changes to "700+ results". Maybe we should move the "show more" button to the end of the list so that you have to scroll down to see it. Any thoughts @rrhyne ?

> Would be awesome to show files in a frecency order rather than lexical (perhaps a flag to enable that). This is a larger feature, I know :)

@tsenart can you give a concrete example that compares the current ordering with frecency ordering? The current ordering displays filenames that have the closest "fuzzy match" according to this npm package (https://github.com/jhawthorn/fzy.js/). How heavily should the recency "match" be weighed against the fuzzy match?

Ideally, we should avoid flags/settings where possible. The default behavior should offer a good enough experience for everybody. 

> Inspired by @tjdevries telescope, I now kind of expect a fuzzy finder to show a preview of the file on the side. That'd be awesome too.

Opening a file should actually be fast since it stays within the React SPA. The "focused item" (the filename you have selected by pressing Up/Down arrow keys) should also persist between blob views, but a recent regression caused the focus item to reset to the top item whenever you open a new blob view.

**Comment by limitedmage** [comment]
> I did notice that the browser keyboard shortcut for reopen tab (cmd + shift + T for mac) is ignored and instead opens this modal, but open a new tab (cmd + T) works fine and opens the tab.

We should only open the panel if the `T` key is pressed with no modifiers.

**Comment by olafurpg** [comment]
@ryanslade I opened https://github.com/sourcegraph/sourcegraph/pull/22133 fixing the issue where files got re-downloaded the second time you activate the fuzzy finder. The files should be cached now within the React single-page app but they get re-downloaded if you open a new tab (or if you switch between `@main` to the resolved git commit URL).

**Comment by olafurpg** [comment]
> @olafurpg We should only open the panel if the T key is pressed with no modifiers.

@limitedmage the keyboard shortcut is only activated when there are no modifiers. The `held` property is undefined

![image](https://user-images.githubusercontent.com/1408093/122274234-a1f43900-cee2-11eb-8b8b-44b1c5d176d1.png)

@coury-clark Thank you for trying it out and sharing your feedback! I am able to reproduce your issue in Firefox but not in Chrome. Are you also using Firefox?

**Comment by coury-clark** [comment]
@olafurpg Sorry, I should have included that. I am on Firefox 89.0.1 (64-bit) and seeing this issue.

**Comment by rrhyne** [comment]
@olafurpg in repsonse to @tsenart's comment on width, we would need to reduce width overall for the entire window, then ensure line breaking looks ok for the list elements. Here's a quick preview that isn't entirely correct (we would want to address lineheight of the wrapped item, vs. margin between elements) but gives you an idea of what that would look like. Tomás, would this solve your issue with the width?

If we like this, a more careful effort than I have time for would analyze the width of file paths in a few repositories so that we may more carefully consider the width of the modal. 

![image](https://user-images.githubusercontent.com/539268/122303767-e4a11b80-ceb8-11eb-8e0b-35b9b459aa12.png)

**Comment by tsenart** [comment]
> @tsenart can you give a concrete example that compares the current ordering with frecency ordering?

I'm talking about the original ordering, before the user inputs anything at all. It's a nice way to quickly jump to recent files when you don't necessarily remember the name. It's definitely a nice to have, not a must.

> Opening a file should actually be fast since it stays within the React SPA. 

Nice! 

> @olafurpg in repsonse to @tsenart's comment on width, we would need to reduce width overall for the entire window, then ensure line breaking looks ok for the list elements

One solution to convert long file paths to shorter ones is to replace long path components with mnemonics, and expand those path components that match the input query.

**Comment by olafurpg** [comment]
Thank you everybody for the feedback! I am now closing this mega-issue in favor of focused issues (see links above). To share more feedback on the fuzzy finder, please post a comment in the discussion here https://github.com/sourcegraph/sourcegraph/discussions/22230. GitHub Discussions support threading and converting threads into issues, which is perfect for this kind of use-case 🙂
